### ISSUE TITLE
code of temporal encoding

### ISSUE THREAD

**Issue by kiki800917** [michaeltrs/DeepSatModels#4]
Dear authors, thanks for giving us this amazing paper. I have a little confused thing abot this following snippet of code:
  
 def forward(self, x):
        x = x.permute(0, 1, 4, 2, 3)
        B, T, C, H, W = x.shape

        xt = x[:, :, -1, 0, 0]
        x = x[:, :, :-1]
        xt = (xt * 365.0001).to(torch.int64)
        xt = F.one_hot(xt, num_classes=366).to(torch.float32)

        xt = xt.reshape(-1, 366)
        temporal_pos_embedding = self.to_temporal_embedding_input(xt).reshape(B, T, self.dim)
        x = self.to_patch_embedding(x)
        x = x.reshape(B, -1, T, self.dim)

this code is on class TSViT(nn.Module) of TSViTdense.py, could you explain what do this code doing? thanks alot!

**Comment by michaeltrs** [comment]
Hi @kiki800917, 

Thanks for your interest in our work and for your kind words. 

This code snippet essentially reformats model inputs (as received by the data pipeline) to be used by TSViT. In more detail:

- xt = x[:, :, -1, 0, 0]
- x = x[:, :, :-1]

The dataloader [concats satellite images with time information (day of year)](https://github.com/michaeltrs/DeepSatModels/blob/b56cccda3be49122e8d9815c7780fa2af1e01757/data/PASTIS24/data_transforms.py#L198) and returns it as a single tensor. This operation essentially separates time from image data, time is the last channel, images are all remaining ones.

- xt = (xt * 365.0001).to(torch.int64)

now day is in range [0, 1], to get it in [0, 365] range we multiply by 365. This [also reverses an operation performed in the dataloader](https://github.com/michaeltrs/DeepSatModels/blob/b56cccda3be49122e8d9815c7780fa2af1e01757/data/PASTIS24/data_transforms.py#L127).

-   xt = F.one_hot(xt, num_classes=366).to(torch.float32)
-   xt = xt.reshape(-1, 366)
-   temporal_pos_embedding = self.to_temporal_embedding_input(xt).reshape(B, T, self.dim)

Now we transform time information into a one-hot representation to effectively use as input to the  self.to_temporal_embedding_input() layer. In the paper we discuss how to use time-specific position encodings by keeping track of all dates in a training set. This can be implemented as a lookup table for these specific dates [as in TSViT_lookup](https://github.com/michaeltrs/DeepSatModels/blob/b56cccda3be49122e8d9815c7780fa2af1e01757/models/TSViT/TSViTdense.py#L439), however, in practice it is much faster to assume a range of dates (here 1 calendar year = 365 days), initialise an embedding for each day P (365 X d) and implement time-specific encodings as matmul(P, xt) where xt is a one-hot encoding of a value in [0, 365]. This way we sub-utilise P as some rows in P are never used (days not found in train data), however, it is much faster to train than TSViT_lookup. This is just an implementation choice. 

-   x = self.to_patch_embedding(x)
-   x = x.reshape(B, -1, T, self.dim)

These operation tokenize images as shown in Figure 3 and reshape to a tensor of time series for all batch samples and locations.
